src/t_stream.c

```diff
@@ -754,93 +769,53 @@ int64_t streamTrim(stream *s, streamAddTrimArgs *args) {
-            }
-
-            int stop;
-            if (trim_strategy == TRIM_STRATEGY_MAXLEN) {
-                stop = s->length <= maxlen;
-            } else {
-                /* Following IDs will definitely be greater because the rax
-                 * tree is sorted, no point of continuing. */
-                stop = streamCompareID(&currid, id) >= 0;
-            }
-            if (stop)
-                break;
-
-            if (flags & STREAM_ITEM_FLAG_SAMEFIELDS) {
-                to_skip = master_fields_count;
-            } else {
-                to_skip = lpGetInteger(p); /* Get num-fields. */
-                p = lpNext(lp,p); /* Skip num-fields. */
-                to_skip *= 2; /* Fields and values. */
-            }
-
-            while(to_skip--) p = lpNext(lp,p); /* Skip the whole entry. */
-            p = lpNext(lp,p); /* Skip the final lp-count field. */
-
-            /* Mark the entry as deleted. */
-            if (!(flags & STREAM_ITEM_FLAG_DELETED)) {
-                intptr_t delta = p - lp;
-                flags |= STREAM_ITEM_FLAG_DELETED;
-                lp = lpReplaceInteger(lp, &pcopy, flags);
-                deleted_from_lp++;
-                s->length--;
-                p = lp + delta;
-            }
-        }
-        deleted += deleted_from_lp;
-
+         * if they are already not deleted and perform a garbage collection
+         * if necessary. */
+        struct trimContext context = {
+            .perform_garbage_collection = 0,
+            .processed_all_trimmed_entries = 0,
+            .master_fields = (uint32_t)master_fields_count,
+            .s = s,
+            .id = id,
+            .master_id = &master_id,
+            .maxlen = maxlen,
+            .trim_strategy = trim_strategy,
+            .num_entries = entries,
+            .num_deleted = num_deleted,
+            .num_deleted_by_trim = 0
+        };
+        lp = lpDeleteRanges(lp, p, listpackGetNextDeletedRangeInXTrim, (void *)&context);
+        deleted += context.num_deleted_by_trim;
+        
         /* Now we update the entries/deleted counters. */
         p = lpFirst(lp);
-        lp = lpReplaceInteger(lp,&p,entries-deleted_from_lp);
+        lp = lpReplaceInteger(lp, &p, context.num_entries);
         p = lpNext(lp,p); /* Skip deleted field. */
-        int64_t marked_deleted = lpGetInteger(p);
-        lp = lpReplaceInteger(lp,&p,marked_deleted+deleted_from_lp);
-        p = lpNext(lp,p); /* Skip num-of-fields in the master entry. */
-
-        /* Here we should perform garbage collection in case at this point
-         * there are too many entries deleted inside the listpack. */
-        entries -= deleted_from_lp;
-        marked_deleted += deleted_from_lp;
-        if (entries + marked_deleted > 10 && marked_deleted > entries/2) {
-            /* TODO: perform a garbage collection. */
+        
+        if (context.perform_garbage_collection) {
+            lp = lpReplaceInteger(lp, &p, 0);
```

Comment:
why setting the fileds num to 0? trim doesn't trim the whole listpack, right?

Reply:
If there are 10 entries, use`XDEL` to delete the 9th one first. At this time, `marked_deleted` is 1. If use xtrim to delete 5 entries and trigger gc, at this time, deleted fields should be 1 instead of 0. right?

Reply:
> If there are 10 entries, useXDEL to delete the 9th one first. At this time, marked_deleted is 1. If use xtrim to delete 5 entries and trigger gc, at this time, deleted fields should be 1 instead of 0. right?

No, in this case, the garbage collection not only physically deletes the first 5 entries but also the 9th entry.